Jameswtruher/travisdailybuild

JamesWTruher opened the pull request:
enable daily test run support in travis

This also includes a number of test changes which need to be reviewed by the test owners
@lzybkr I've made changes in parser tests to support async execution to avoid test run hangs which have sporadically appeared. Now, if one of those tests hang, it (the test) will fail rather than the entire run.
@Francisco-Gamino, I've made changes to the help tests, counter and management tests

once this is merged, I'll have an addition PR to put the test run badge in the ReadMe.md (after I create the cron job to actually run the tests)

JamesWTruher committed:
Stifle progress output in build.psm1 for some operations

Modify test failure presentation to use platform available XML methods

JamesWTruher committed:
Add timeout support for returning runtime parsing errors

Some of the language/parser tests have been hanging in a non-reproducable manner which
causes the CI system to invalidate the entire run. This change adds support for timeout
which will fail a test if it runs to long, rather than invalidate the entire run.

current behavior is still supported, and is not done in a new session:
PS> get-runtimeerror -src '1/'
At line:1 char:3
+ 1/
+   ~
You must provide a value expression following the '/' operator.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : ExpectedValueExpression

Adding a timeout will do the operation in a async powershell session
PS> get-runtimeerror -src '1/' -timeout 5
You must provide a value expression following the '/' operator.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : ExpectedValueExpression

If the operation takes longer than the supplied timeout, a timeout error will be returned
PS> get-runtimeerror -src 'start-sleep 6' -timeout 2
get-runtimeerror : Operation Timed Out ('start-sleep 6')
At line:1 char:1
+ get-runtimeerror -src 'start-sleep 6' -timeout 2
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Get-RuntimeError

JamesWTruher committed:
Modify native linux command tests to skip on Windows and pending on Mac

JamesWTruher committed:
remove verbose and progress output from help tests

JamesWTruher committed:
Be sure that Feature Counter tests only run on Windows

Also, only call add-type in CounterTestHelperFunctions.ps1 if we're going to actually run the tests

JamesWTruher committed:
do not run any get-computerinfo tests on non-windows systems

JamesWTruher committed:
suppress progress output from PowerShell Get tests

JamesWTruher committed:
remove -quiet from API and CRON Builds

Travis watches output from the build to ensure that it hasn't hung
we need to find a balance between too much output and not enough output.
A run which has too much output is killed because it looks like an error loop
A run which has too little output is killed because it looks like a hang

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Import-Counter.Tests.ps1:
```diff
@@ -110,7 +118,7 @@ function RunTest($testCase)
 
         $cmd = ConstructCommand $testCase
-        Write-Host "Command to run: $cmd"
+        # Write-Host "Command to run: $cmd"
```
Should just delete this line.

JamesWTruher replied:
done!

adityapatwardhan commented on test/powershell/Modules/PowerShellGet/PowerShellGet.Tests.ps1:
```diff
@@ -90,6 +93,7 @@ function Remove-InstalledModules
     Get-InstalledModule -Name $ContosoServer -AllVersions -ErrorAction SilentlyContinue | Uninstall-Module -Force
 }
 
+
```
Extra line.

JamesWTruher replied:
deleted

adityapatwardhan commented on build.psm1:
```diff
@@ -2813,6 +2821,8 @@ function Restore-PSModule
     log ("Name='{0}', Destination='{1}', Repository='{2}'" -f ($Name -join ','), $Destination, $RepositoryName)
 
+    # do not output progress
+    $ProgressPreference = "SilentlyContinue"
```
Do we need to reset this back to previous setting when we complete the script?

JamesWTruher replied:
nope - test tests are executed in a new scope, so it gets set back when the script exits

JamesWTruher committed:
Remove commented line in Import-Counter.Tests.ps1

Remove extraneous extra line in PowerShellGet.Tests.ps1

lzybkr commented on test/powershell/Language/LanguageTestSupport.psm1:
```diff
@@ -24,18 +24,57 @@ function Get-RuntimeError
     }
     catch 
     {
+        write-verbose -verbose "caught"
```
Is this a left-over debugging message?
`caught` isn't a very useful message by itself, and I don't see any other `Write-Verbose` calls in this function to give it context.

JamesWTruher replied:
indeed yes - pulled

---
In reply to: [94878328](https://github.com/PowerShell/PowerShell/pull/2958#discussion_r94878328) [](ancestors = 94878328)

lzybkr commented on test/powershell/Language/LanguageTestSupport.psm1:
```diff
@@ -76,7 +115,7 @@ function ShouldBeParseError
         {
             It "error should happen at parse time, not at runtime" -Skip {}
-            $errors = Get-RuntimeError -Src $src
+            $errors = Get-RuntimeError -Src $src -Timeout 5
```
I'd feel a bit better if we waited a bit longer, maybe 10 seconds - in case we're on a really slow VM or we add a test that is somewhat longish (like 1s or so).

JamesWTruher replied:
it was SWAG for sure, I don't have trouble with having this be 10

---
In reply to: [94878763](https://github.com/PowerShell/PowerShell/pull/2958#discussion_r94878763) [](ancestors = 94878763)

lzybkr commented on build.psm1:
```diff
@@ -835,7 +836,14 @@ function Test-PSPesterResults
         logerror "TEST FAILURES"
-        foreach ( $testfail in $x.SelectNodes('.//test-case[@result = "Failure"]'))
+        # switch between methods, SelectNode is not available on dotnet core
+        if ( "System.Xml.XmlDocumentXPathExtensions" -as "Type" ) {
```
As a general rule, you should prefer `-as [Type]` instead of `-as "Type"` because the former resolves the type just once while the later resolves it every time it is executed.

JamesWTruher replied:
changed

---
In reply to: [94879724](https://github.com/PowerShell/PowerShell/pull/2958#discussion_r94879724) [](ancestors = 94879724)

lzybkr reviewed (commented):
Help me understand the timeout on the parser tests - the test will still fail, but just fail faster?  Or something else?

I wonder how hard it would be to get a core dump when we see the timeout - upload that as an artifact, and debug later.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Import-Counter.Tests.ps1:
```diff
@@ -35,12 +43,12 @@ function SetScriptVars([string]$rootPath, [int]$maxSamples, [bool]$export)
 
     $script:counterSamples = $null
-    if ($maxSamples)
+    if ($maxSamples -and $IsWindows)
```
Seems `Get-Counter` don't work on IoT.
`SetScriptVars` is only called from BeforeAll blocks so we can mask it there by `SkipCounterTests`.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Import-Counter.Tests.ps1:
```diff
@@ -35,12 +43,12 @@ function SetScriptVars([string]$rootPath, [int]$maxSamples, [bool]$export)
     }
 
-    if ($export)
+    if ($export -and $IsWindows)
```
The same about IoT.

iSazonov commented on test/powershell/Language/LanguageTestSupport.psm1:
```diff
@@ -24,18 +24,57 @@ function Get-RuntimeError
+            $ps.AddScript($src) > $null
+            $ar = $ps.BeginInvoke()
+            # give it 250 milliseconds to complete
+            start-sleep -mill 250
```
Parser tests is very fast (some ms). This change will slow them down in many times.
And how will it help us to find the root of the problem? Travis already has a global monitoring of a job duration and automatically kills a hung job. Maybe we ask them to take a dump in this moment?

JamesWTruher replied:
yes, it will slow down some tests, but I reckoned that it was better to slow down the test than to invalidate an entire run - that's what I'm trying to avoid. When running a full test pass I saw nearly 100% test runs killed for lack of activity. Each one of those runs were hung in the language tests.

---
In reply to: [94967709](https://github.com/PowerShell/PowerShell/pull/2958#discussion_r94967709) [](ancestors = 94967709)

iSazonov replied:
I am concerned that we conceal the problem. Is there a way to reproduce this problem except to get a hung _accidentally_? I believe that we in the allowed time could identify the problem and fix it by analyzing a dump without slow down tests.

JamesWTruher replied:
while I agree that we need to determine root cause, i disagree about where. Catching this in CI has been extremely difficult and collecting dumps in Travis is really something I would rather avoid (if it's even possible).
we need to track down this issue, but I don't believe CI is the appropriate place

---
In reply to: [94993680](https://github.com/PowerShell/PowerShell/pull/2958#discussion_r94993680) [](ancestors = 94993680)

JamesWTruher replied:
an issue has already been created for the hanging behavior. #2748

---
In reply to: [95024875](https://github.com/PowerShell/PowerShell/pull/2958#discussion_r95024875) [](ancestors = 95024875)

JamesWTruher committed:
Change `-as "type"` to `-as [type]` in build.psm1

Alter timeout to 10 seconds to be improve chances of not timing out for runtime parser checks
improve logic for counter tests to also skip for IoT

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Import-Counter.Tests.ps1:
```diff
@@ -7,7 +7,17 @@ $cmdletName = "Import-Counter"
+    $ShouldRun = $false
+}
+
+if ( $ShouldRun )
```
`$ShouldRun` looks like duplication of `SkipCounterTests`

JamesWTruher commented:
@lzybkr I don't seem to be able to follow up on your question about the timeout in place.

Currently, when one of the parser tests hang, the entire test run is killed after about 10 minutes of inactivity. The timeout code enables us to fail a single test and have the rest of the run continue. I agree that capturing a dump at that point would be fantastic, but I don't know how to do that. Can you provide instructions? I'll be very happy to add it.

JamesWTruher commented:
@TravisEz13 an issue has already been created for the hanging behavior. https://github.com/PowerShell/PowerShell/issues/2748

lzybkr commented:
I don't know, but a search of 'create linux core dump' suggests that `gcore` or `kill` may be capable of creating the core dump.

iSazonov commented:
It seems it is impossible that _hung_ process killed himself and unloaded a dump.
As I mentioned above the best way to get the dump is to ask Travis team add the new option to travis.yml They already control session timeout and may at this point easily take a dump and upload it in the specified (in travis.yml) location. This can be useful in the future to solve other problems.

Possible workarround http://jsteemann.github.io/blog/2014/10/30/getting-core-dumps-of-failed-travisci-builds/

lzybkr commented:
We don't know if the process is hung, but clearly a thread is hung, so Jim's change may help in creating the dump. But I agree the CI system should provide a way to debug of they don't already. Appveyor definitely does provide rdp access, maybe Travis already has something similar.

iSazonov commented:
I found nothing in Travis help but I discovered long discussion about travis dumps travis-ci/travis-ci#3754

JamesWTruher commented:
will that provide you with what you need, given managed code? As you're the author/owner of these particular tests, shall I just open an issue for you to track?

---
In reply to: [271043086](https://github.com/PowerShell/PowerShell/pull/2958#issuecomment-271043086) [](ancestors = 271043086)

JamesWTruher committed:
use the existing function of SkipCounterTests rather than duplicate the logic in import-counter.tests.ps1

JamesWTruher commented:
@Francisco-Gamino, @TravisEz13 my last commit should have addressed your request

lzybkr commented:
@JamesWTruher

If I'm reading the logs correctly, the async changes didn't help at all, and most likely are a net negative right now.  See these logs:

https://travis-ci.org/PowerShell/PowerShell/jobs/195355364

Notice 10 minutes with no output - so we failed to kill the async test and did not get further results.
Also notice the error tests take ~250ms when they should really take about ~1ms or less.

iSazonov commented:
The same https://travis-ci.org/PowerShell/PowerShell/jobs/195856711
